### ISSUE TITLE
Trying to set-up plugin deployment with Daxif

### ISSUE THREAD

**Issue by robertraaijmakers** [delegateas/Daxif#17]
Hi there,

I've seen your presentation at eXtreme and was trying to remove some of the deadly sins of CRM development. I've removed one already using the early bound and was trying to set-up automated deployment for a new plugin project. But I'm encountering some issue and can't figure out what is wrong with my set-up.

I will try to explain the set-up without giving to much customer details.

- Simple plugin solution, with 3 plugins on a Dynamics 365 V9.0 environment using the Dynamics 365 9.0.0.7 Microsoft.Xrm.Sdk (nuget).
- Running the tool from Visual Studio 2017 via F# Interactive.
- The plugins aren't registered in Dynamics yet.

**Error message**
(2018-04-10 15:25:20Z) [Info]: Environment: "XXXX DEV" (https://xxx.crm4.dynamics.com/XRMServices/2011/Organization.svc)
(2018-04-10 15:25:20Z) [Info]: User: xxx@xxx.onmicrosoft.com
(2018-04-10 15:25:21Z) [Info]: DAXIF# v.4.1.1.0
(2018-04-10 15:25:21Z) [Info]: Action: Plugin synchronization
(2018-04-10 15:25:21Z) [Info]: Comparing plugins registered in CRM versus those found in your local code
(2018-04-10 15:25:21Z) [Verbose]: Loading local assembly and it's plugins
Binding session to 'D:\Projects\xxx\xxxx\xxxx.Plugins\xxx.Plugins.Tools\Daxif\bin\Microsoft.Xrm.Sdk.dll'...
System.Exception: Failed to fetch plugin configuration from plugin assembly. This error can be caused if an old version of Plugin.cs is used.\nFull Exception: Unable to load one or more of the requested types. Retrieve the LoaderExceptions property for more information.
   at Microsoft.FSharp.Core.PrintfModule.PrintFormatToStringThenFail@1645.Invoke(String message)
   at DG.Daxif.Modules.Plugin.PluginDetection.getPluginsFromAssembly(Assembly asm)
   at DG.Daxif.Modules.Plugin.PluginDetection.getAssemblyContextFromDll(String projectPath, String dllPath, PluginIsolationMode isolationMode, Boolean ignoreOutdatedAssembly)
   at DG.Daxif.Modules.Plugin.MainHelper.loadAndValidateAssembly[a](FSharpFunc`2 proxyGen, String projectPath, String dllPath, PluginIsolationMode isolationMode, Boolean ignoreOutdatedAssembly)
   at DG.Daxif.Modules.Plugin.MainHelper.analyze[a](FSharpFunc`2 proxyGen, String projectPath, String dllPath, String solutionName, PluginIsolationMode isolationMode, Boolean ignoreOutdatedAssembly)
   at DG.Daxif.Modules.Plugin.Main.syncSolution[a](FSharpFunc`2 proxyGen, String projectPath, String dllPath, String solutionName, PluginIsolationMode isolationMode, Boolean ignoreOutdatedAssembly, Boolean dryRun)
   at DG.Daxif.Plugin.Sync(Environment env, String assemblyPath, String projectPath, String solutionName, FSharpOption`1 dryRun, FSharpOption`1 isolationMode, FSharpOption`1 ignoreOutdatedAssembly, FSharpOption`1 logLevel)
   at <StartupCode$FSI_0074>.$FSI_0074.main@() in D:\Projects\xxx\xxx\xxx.Dynamics365.Plugins\xxx.Dynamics365.Plugins.Tools\Daxif\PluginSyncDev.fsx:line 12
Stopped due to error

**Plugin Definition**
```
public class AddAccountToAccessTeam : Plugin
    {
        /// <summary>
        /// Initializes a new instance of the <see cref="AddAccountToAccessTeam"/> class.
        /// And registers the plugin on the correct plugin events.
        /// </summary>
        public AddAccountToAccessTeam()
            : base(typeof(AddAccountToAccessTeam))
        {
            RegisterPluginStep<account>(EventOperation.Update, ExecutionStage.PostOperation, ExecuteAddAccountToAccessTeam)
                .SetExecutionMode(ExecutionMode.Synchronous);
        }

private void ExecuteAddAccountToAccessTeam(LocalPluginContext localContext)
        {
        }
}
```

**Plugin Class used**
I've tried to use the following two classes
- https://github.com/TomMalow/ExtendedPlugin.cs
- https://github.com/delegateas/Daxif/blob/master/src/Delegate.Daxif/Resources/Plugin.cs

**Dynamcs DLL**
I've also tried to replace the \Daxif\bin\Microsoft.Xrm.Sdk.dll to use the latest Xrm.Sdk (v 9.0.0.7). But that gives me the same error.

I'm a bit lost. Is there a sample somewhere with a very basic set-up that works?

Thanks!

**Comment by henrikhannemose** [comment]
Great to hear that you managed to get early bound up and running! It is also nice to hear that you are looking into automating your plugin registration.

I am not sure I can spot exactly the error you are experiencing, but I have some things that jump out at me that I would like us to look closer at:

**The casing of account:**
Why is `account` lower case in `RegisterPluginStep<account>`? Which class does that refer to? I would expect the generated `Account` class from XrmContext to have a different capitalization.

**The output from Plugin.Sync**
When calling `Plugin.Sync(Env.dev, pluginDll, pluginProjFile, SolutionInfo.name)`: What is your `pluginDll`? Are you pointing at an ILMerged DLL?

**Plugin Class used**
This one should be the one to use: https://github.com/delegateas/Daxif/blob/master/src/Delegate.Daxif/Resources/Plugin.cs

**Dynamics DLL**
You don't need to change the DLL version. DAXIF is able to connect to v9-environments using the 8.2 DLL it ships with.

**Comment by robertraaijmakers** [comment]
Thanks for the quick response!

The casing of account: it was a typo in my copy/paste action and is indeed upper case Account as class.
The output from Plugin.Sync, the plugin dll is the normal dll generated by my visual studio project (just project name .dll) not a ILMerged dll (I don't need to merge external dlls, just using the out of the box functionality, so no need to ILMerge stuff (I suppose, it works with the plugin registration toolkit)).

I updated and re-ran with the plugin class that you provided. And the same error exists.

I've done some additional testing myself. And recreated a complete new plugin solution from scratch with no plugins and then register that solution in Dynamics with Daxif. That all works fine (for a 9.0 environment). So after that I was curious what was breaking the Daxif plugin sync and started adding the plugins one by one. It looks like that when I add resource files (.resx) to my solution, the inspection fails and gives the error as described in the opening post.

**Comment by henrikhannemose** [comment]
Thanks for digging deeper into this. I am afraid I cannot reproduce the issue by adding a .resx file to my project:
![.resx file in plugin project](https://user-images.githubusercontent.com/1955579/38622960-b7d8f8c8-3da4-11e8-8504-c3772882a327.png)

`Plugin.Sync` still succeeds for me with the above .resx file in my project. Are you able to share a minimal example of this error with the .resx file?

**Comment by robertraaijmakers** [comment]
Yes, I will try to send you the solution. I found the file that causes the issue. It wasn't the resource file as I thought. It has to do with a file that uses the following usings

using xxxxx.Dynamics365.Plugins.Models.XrmContext;
using Microsoft.Crm.Sdk.Messages;
using Microsoft.Xrm.Sdk;
using System;
using System.ServiceModel;

Can you pm me an email address, then I can try to provide a minimal solution so that you can try to reproduce the error.

**Comment by robertraaijmakers** [comment]
[Conferencing.Dynamics365.Plugins.zip](https://github.com/delegateas/Daxif/files/1899359/Conferencing.Dynamics365.Plugins.zip)

I removed all customer references and stripped the solutions. Hope this will help.

**Comment by robertraaijmakers** [comment]
[Conferencing.Dynamics365.Plugins.zip](https://github.com/delegateas/Daxif/files/1899366/Conferencing.Dynamics365.Plugins.zip)

I stripped the solution and removed the customer references. Hope this helps.

**Comment by robertraaijmakers** [comment]
Thanks for looking into it! Is there a way I can try to help to resolve this issue?

**Comment by henrikhannemose** [comment]
tl;dr: Try adding `#r @"bin\Microsoft.Crm.Sdk.Proxy.dll"` to `_Config.fsx`.

---
I have now taken a closer look at this. 

The MCVE above uses stuff from the `Microsoft.Crm.Sdk.Messages`-namespace. This namespace lives in the [`Microsoft.Crm.Sdk.Proxy.dll`](https://msdn.microsoft.com/en-us/library/gg334621.aspx). I have now added some more verbosity to DAXIF (in 6dc01be), such that it outputs a message similar to the following when it is unable to load DLLs:
```
[Info]: Error when loading the following assemblies (they will be skipped):
[Info]: Could not load file or assembly 'Microsoft.Crm.Sdk.Proxy, Version=9.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
```
Additionally, DAXIF will attempt to continue. In this case that just results in a failure at a later step. 
I can see (with Process Monitor) that fsi.exe attempts to load `Microsoft.Crm.Sdk.Proxy.dll` from `Daxif/Microsoft.Crm.Sdk.Proxy.dll` (not from `Daxif/bin/Microsoft.Crm.Sdk.Proxy.dll` where it is located in all DAXIF installations).

A fix is to make sure that the DLL is loaded from the right location. This can be done by adding the following to e.g. `_Config.fsx`:
`#r @"bin\Microsoft.Crm.Sdk.Proxy.dll"`
I have added this by default in `_Config.fsx`, but it has not yet been released on NuGet.

**Comment by robertraaijmakers** [comment]
Sweet, works like a charm! Thanks!

**Comment by henrikhannemose** [comment]
That's great to hear. Feel free to open new issues if you encounter other problems along the way - and thanks for trying out the tools! :)
